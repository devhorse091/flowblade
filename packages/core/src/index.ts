export {
  assertQueryResultError,
  assertQueryResultSuccess,
  isQueryResultError,
  isQueryResultSuccess,
} from './data-source/assert';
export type {
  DatasourceInterface,
  DatasourceQueryInfo,
} from './data-source/datasource.interface';
export type {
  AsyncQueryResult,
  InferAsyncQueryResultData,
  InferQueryResultData,
  QueryResult,
  QueryResultError,
  QueryResultMeta,
  QueryResultSuccess,
} from './data-source/query-result';
export {
  createResultError,
  createResultSuccess,
} from './data-source/query-result-factories';
export {
  type PerformanceSpan,
  type QError,
  QMeta,
  QResult,
} from './query-result/q-result';
export { AsyncResult, Result } from 'typescript-result';
