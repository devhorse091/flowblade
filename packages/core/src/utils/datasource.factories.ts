import type { QMeta, QMetaSqlSpan } from '../meta/q-meta';
import { type QError, QResult } from '../query-result/q-result';

/**
 * Create a new SQL span with sql query + params and initial timeMs: 0, affectedRows: 0.
 */
export const createSqlSpan = ({
  sql,
  params,
  timeMs,
  affectedRows,
}: {
  sql: QMetaSqlSpan['sql'];
  params: QMetaSqlSpan['params'];
  timeMs?: number;
  affectedRows?: number;
}): QMetaSqlSpan => {
  return {
    type: 'sql',
    sql,
    params,
    timeMs: timeMs ?? 0,
    affectedRows: affectedRows ?? 0,
  };
};

export const createQResultSuccess = <T extends unknown[]>(
  data: T,
  meta: QMeta
): QResult<T, never> => {
  return new QResult({
    meta,
    data,
  });
};

export const createQResultError = (
  error: QError,
  meta: QMeta
): QResult<never, QError> => {
  return new QResult({
    meta,
    error,
  });
};
