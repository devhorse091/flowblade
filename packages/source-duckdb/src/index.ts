export * from './datasource/duckdb-async-datasource';
