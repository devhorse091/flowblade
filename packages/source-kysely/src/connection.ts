import { MssqlDialect } from 'kysely';
import * as tarn from 'tarn';
import * as tedious from 'tedious';

export const createDialect = (
  tediousConfig: tedious.ConnectionConfiguration
) => {
  return new MssqlDialect({
    tarn: {
      ...tarn,
      options: {
        min: 0,
        max: 10,
      },
      /**
       * @todo when https://github.com/kysely-org/kysely/pull/1073/files is merged
       * Controls whether connections are validated before being acquired from the pool. Connection validation performs additional requests to the database.
       *
       * Defaults to `true`.
       */
      // validateConnections: false
    },
    tedious: {
      ...tedious,
      // See https://github.com/kysely-org/kysely/issues/1161#issuecomment-2384539764
      TYPES: { ...tedious.TYPES, NVarChar: tedious.TYPES.VarChar },
      connectionFactory: () => {
        return new tedious.Connection(tediousConfig);
      },
      /**
       * @todo when https://github.com/kysely-org/kysely/pull/1073/files is merged
       *
       * Controls whether connections are reset to their initial states when released back to the pool. Resetting a connection performs additional requests to the database.
       * See {@link https://tediousjs.github.io/tedious/api-connection.html#function_reset | connection.reset}.
       *
       * Defaults to `true`.
       */
      // resetConnectionOnRelease: false
    },
  });
};
