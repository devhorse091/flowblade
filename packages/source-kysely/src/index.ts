export * from './datasource/kysely-datasource';
export * from './helpers/datasource-helpers';
export { createKyselySqlServerDialect } from './utils/create-kysely-sql-server-dialect';
export { TediousConnUtils } from './utils/tedious-conn-utils';
