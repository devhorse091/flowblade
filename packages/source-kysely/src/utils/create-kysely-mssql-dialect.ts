import { MssqlDialect } from 'kysely';
import * as tarn from 'tarn';
import * as tedious from 'tedious';

/**
 * Create a Kysely dialect for Microsoft SQL Server.
 *
 * @example
 * ```typescript
 * const jdbcDsn = "sqlserver://localhost:1433;database=db;user=sa;password=pwd;trustServerCertificate=true;encrypt=false";
 * const tediousConfig = TediousConnUtils.fromJdbcDsn(jdbcDsn);
 * const tediousConnection = new Tedious.Connection(tediousConfig);
 *
 * const dialect = createKyselyMssqlDialect(tediousConfig, {
 *   // Optional tarn pool options
 *   tarnPool: {
 *     min: 0,
 *     max: 10
 *   }
 * });
 *
 * const db = new Kysely<DB>({
 *   dialect
 * })
 * ```
 */
export const createKyselyMssqlDialect = (
  tediousConfig: tedious.ConnectionConfiguration,
  options?: {
    tarnPool?: {
      /** default: 0 */
      min?: number;
      /** default: 10 */
      max?: number;
    };
  }
) => {
  const { tarnPool = {} } = options ?? {};
  const { min = 0, max = 10 } = tarnPool;
  return new MssqlDialect({
    tarn: {
      ...tarn,
      options: {
        min,
        max,
      },
      /**
       * @todo when https://github.com/kysely-org/kysely/pull/1073/files is merged
       * Controls whether connections are validated before being acquired from the pool. Connection validation performs additional requests to the database.
       *
       * Defaults to `true`.
       */
      // validateConnections: false
    },
    tedious: {
      ...tedious,
      // See https://github.com/kysely-org/kysely/issues/1161#issuecomment-2384539764
      TYPES: { ...tedious.TYPES, NVarChar: tedious.TYPES.VarChar },
      connectionFactory: () => {
        return new tedious.Connection(tediousConfig);
      },
      /**
       * @todo when https://github.com/kysely-org/kysely/pull/1073/files is merged
       *
       * Controls whether connections are reset to their initial states when released back to the pool. Resetting a connection performs additional requests to the database.
       * See {@link https://tediousjs.github.io/tedious/api-connection.html#function_reset | connection.reset}.
       *
       * Defaults to `true`.
       */
      // resetConnectionOnRelease: false
    },
  });
};
