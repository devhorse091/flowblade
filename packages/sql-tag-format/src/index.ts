export type { SqlFormatterDialect, SqlFormatterOptions } from './sql-formatter';
export { SqlFormatter } from './sql-formatter';
