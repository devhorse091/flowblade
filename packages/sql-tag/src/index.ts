export { sql } from './sql';
export type { TaggedSql } from './types';
