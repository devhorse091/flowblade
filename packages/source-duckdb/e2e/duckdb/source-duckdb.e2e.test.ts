import { sql } from '@flowblade/sql-tag';
import { afterEach, beforeEach, expectTypeOf } from 'vitest';

import { DuckdbDatasource } from '../../src';
import { createDuckDBE2EMemoryDb } from '../utils/create-duckdb-e2e-memory-db';

describe('DuckDBAsyncDatasource e2e', async () => {
  let ds: DuckdbDatasource;

  beforeEach(async () => {
    const duckdb = await createDuckDBE2EMemoryDb();
    ds = new DuckdbDatasource({ connection: duckdb });
  });

  afterEach(async () => {
    ds.getConnection().close();
  });

  describe('query', () => {
    const params = {
      min: 10,
      max: 99,
      name: 'test',
      createdAt: new Date().toISOString(),
    };

    type Row = { id: number; name: 'test'; createdAt: Date };

    const rawSql = sql<Row>`

      WITH products(productId, createdAt)
          AS MATERIALIZED (
               FROM RANGE(1,100) SELECT 
               range::INT,
               TIMESTAMPTZ '2025-01-01 12:30:00.123456789+01:00'
          )
      
      SELECT productId, 
             ${params.name} as name,
             createdAt
             
      FROM products 
      WHERE productId BETWEEN ${params.min}::INTEGER AND ${params.max}::INTEGER
      AND createdAt < ${params.createdAt}::TIMESTAMPTZ
    `;

    it('should return expected data', async () => {
      const result = await ds.query(rawSql);

      const { data, error } = result;

      expect(error).toBeUndefined();
      expect(data!.length).toBe(90);
      expect(data).toMatchSnapshot();
      expectTypeOf(data!).toEqualTypeOf<Row[]>();
    });

    it('should return expected meta', async () => {
      const result = await ds.query(rawSql);
      const { meta } = result;
      expect(meta.getSpans().length).toBe(1);
      expect(meta.getSpans()).toMatchSnapshot();
    });
  });
});
