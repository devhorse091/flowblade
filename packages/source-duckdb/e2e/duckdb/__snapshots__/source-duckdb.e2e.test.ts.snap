// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DuckDBAsyncDatasource e2e > query > should return expected data 1`] = `
[
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 10,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 11,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 12,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 13,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 14,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 15,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 16,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 17,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 18,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 19,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 20,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 21,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 22,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 23,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 24,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 25,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 26,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 27,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 28,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 29,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 30,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 31,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 32,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 33,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 34,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 35,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 36,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 37,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 38,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 39,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 40,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 41,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 42,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 43,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 44,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 45,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 46,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 47,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 48,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 49,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 50,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 51,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 52,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 53,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 54,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 55,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 56,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 57,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 58,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 59,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 60,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 61,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 62,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 63,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 64,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 65,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 66,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 67,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 68,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 69,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 70,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 71,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 72,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 73,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 74,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 75,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 76,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 77,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 78,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 79,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 80,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 81,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 82,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 83,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 84,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 85,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 86,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 87,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 88,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 89,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 90,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 91,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 92,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 93,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 94,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 95,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 96,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 97,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 98,
  },
  {
    "createdAt": "2025-01-01 12:30:00.123456+01",
    "name": "test",
    "productId": 99,
  },
]
`;

exports[`DuckDBAsyncDatasource e2e > query > should return expected meta 1`] = `
[
  {
    "affectedRows": 90,
    "params": [
      "test",
      10,
      99,
      "2025-01-22T23:22:59.089Z",
    ],
    "sql": "

      WITH products(productId, createdAt)
          AS MATERIALIZED (
               FROM RANGE(1,100) SELECT 
               range::INT,
               TIMESTAMPTZ '2025-01-01 12:30:00.123456789+01:00'
          )
      
      SELECT productId, 
             $1 as name,
             createdAt
             
      FROM products 
      WHERE productId BETWEEN $2::INTEGER AND $3::INTEGER
      AND createdAt < $4::TIMESTAMPTZ
    ",
    "timeMs": 1.6006849999999986,
    "type": "sql",
  },
]
`;
